Import modules

In [304]:
import pandas as pd
import numpy as np
import requests
import matplotlib

Change configuration in pandas so all columns of a dataframe are displayed.

In [305]:
pd.options.display.max_columns = 100
pd.set_option('display.max_rows', None)

Read in api from open5e.com.
This contains information on all spells from Dunegeons & Dragons.
Data source: https://open5e.com/

In [306]:
r = requests.get("https://api.open5e.com/spells/?limit=1000")
r2 = r.json()['results']
open5e = pd.DataFrame(data=r2)
open5e.head()

slug                  name  \
0  abhorrent-apparition  Abhorrent Apparition   
1            accelerate            Accelerate   
2            acid-arrow            Acid Arrow   
3             acid-gate             Acid Gate   
4             acid-rain             Acid Rain   

                                                desc  \
0  You imbue a terrifying visage onto a gourd and...   
1  Choose up to three willing creatures within ra...   
2  A shimmering green arrow streaks toward a targ...   
3  You create a portal of swirling, acidic green ...   
4  You unleash a storm of swirling acid in a cyli...   

                                        higher_level     page     range  \
0  If you cast this spell using a spell slot of 5...            60 feet   
1  When you cast this spell using a spell slot of...              Touch   
2  When you cast this spell using a spell slot of...  phb 259   90 feet   
3  When you cast this spell using a spell slot of...            60 feet   
4  When you cast this spell using a spell slot of...           150 feet   

  components                                           material ritual  \
0          M                   a gourd with a face carved on it     no   
1    V, S, M                                          a toy top     no   
2    V, S, M      Powdered rhubarb leaf and an adder's stomach.     no   
3    V, S, M  a vial of acid and a polished silver mirror wo...     no   
4    V, S, M                                     a drop of acid     no   

         duration concentration casting_time      level  level_int  \
0   Instantaneous            no     1 action  4th-level          4   
1  Up to 1 minute           yes     1 action  3rd-level          3   
2   Instantaneous            no     1 action  2nd-level          2   
3  Up to 1 minute           yes     1 action  7th-level          7   
4  Up to 1 minute           yes     1 action  5th-level          5   

          school                              dnd_class     archetype circles  \
0       illusion                 Bard, Sorcerer, Wizard                         
1  transmutation  Bard, Cleric, Druid, Sorcerer, Wizard                         
2      Evocation                          Druid, Wizard  Druid: Swamp   Swamp   
3    conjuration                                                                
4    conjuration              Sorcerer, Warlock, Wizard                         

  document__slug             document__title    document__license_url  
0           dmag  Deep Magic for 5th Edition  http://open5e.com/legal  
1           dmag  Deep Magic for 5th Edition  http://open5e.com/legal  
2       wotc-srd  Systems Reference Document  http://open5e.com/legal  
3           dmag  Deep Magic for 5th Edition  http://open5e.com/legal  
4           dmag  Deep Magic for 5th Edition  http://open5e.com/legal

A count of all spells in the "open5e" dataframe. This contains more spells than I was expecting.

In [307]:
open5e.count()

slug                     835
name                     835
desc                     835
higher_level             835
page                     835
range                    835
components               835
material                 835
ritual                   835
duration                 835
concentration            835
casting_time             835
level                    835
level_int                835
school                   835
dnd_class                835
archetype                835
circles                  835
document__slug           835
document__title          835
document__license_url    835
dtype: int64

Dungeons & Dragons has had additional spells added to the game from various sources. For the purposes of this prject, I only want information from the "Player's Handbook," referenced in this data as "Systems Reference Document."

Here, I am looking for the titles of all source books.

In [308]:
print(open5e["document__title"].unique())

['Deep Magic for 5th Edition' 'Systems Reference Document' 'Open5e OGL']


Removing all source books other than the "Player's Handbook."

In [309]:
open5e = open5e.loc[open5e['document__title'] == 'Systems Reference Document']
open5e.head()

slug         name  \
2    acid-arrow   Acid Arrow   
5   acid-splash  Acid Splash   
9           aid          Aid   
10        alarm        Alarm   
16   alter-self   Alter Self   

                                                 desc  \
2   A shimmering green arrow streaks toward a targ...   
5   You hurl a bubble of acid. Choose one creature...   
9   Your spell bolsters your allies with toughness...   
10  You set an alarm against unwanted intrusion. C...   
16  You assume a different form. When you cast the...   

                                         higher_level     page    range  \
2   When you cast this spell using a spell slot of...  phb 259  90 feet   
5                                                      phb 211  60 feet   
9   When you cast this spell using a spell slot of...  phb 211  30 feet   
10                                                     phb 211  30 feet   
16                                                     phb 211     Self   

   components                                       material ritual  \
2     V, S, M  Powdered rhubarb leaf and an adder's stomach.     no   
5        V, S                                                    no   
9     V, S, M                   A tiny strip of white cloth.     no   
10    V, S, M   A tiny bell and a piece of fine silver wire.    yes   
16       V, S                                                    no   

         duration concentration casting_time      level  level_int  \
2   Instantaneous            no     1 action  2nd-level          2   
5   Instantaneous            no     1 action    Cantrip          0   
9         8 hours            no     1 action  2nd-level          2   
10        8 hours            no     1 minute  1st-level          1   
16   Up to 1 hour           yes     1 action  2nd-level          2   

           school                      dnd_class     archetype circles  \
2       Evocation                  Druid, Wizard  Druid: Swamp   Swamp   
5     Conjuration               Sorcerer, Wizard                         
9      Abjuration                Cleric, Paladin                         
10     Abjuration  Ranger, Ritual Caster, Wizard                         
16  Transmutation               Sorcerer, Wizard                         

   document__slug             document__title    document__license_url  
2        wotc-srd  Systems Reference Document  http://open5e.com/legal  
5        wotc-srd  Systems Reference Document  http://open5e.com/legal  
9        wotc-srd  Systems Reference Document  http://open5e.com/legal  
10       wotc-srd  Systems Reference Document  http://open5e.com/legal  
16       wotc-srd  Systems Reference Document  http://open5e.com/legal

Counting the entries again.

In [310]:
open5e.count()

slug                     319
name                     319
desc                     319
higher_level             319
page                     319
range                    319
components               319
material                 319
ritual                   319
duration                 319
concentration            319
casting_time             319
level                    319
level_int                319
school                   319
dnd_class                319
archetype                319
circles                  319
document__slug           319
document__title          319
document__license_url    319
dtype: int64

Listing all columns in the open5e dataframe.

In [311]:
open5e.reset_index(inplace=True)

In [312]:
open5e.columns

Index(['index', 'slug', 'name', 'desc', 'higher_level', 'page', 'range',
       'components', 'material', 'ritual', 'duration', 'concentration',
       'casting_time', 'level', 'level_int', 'school', 'dnd_class',
       'archetype', 'circles', 'document__slug', 'document__title',
       'document__license_url'],
      dtype='object')

Removing unneeded columns and columns that will be redundant with later data, and renaming some other columns to suit my preference.

In [313]:
open5e.drop(columns=['index','page', 'level', 'material', 'archetype', 'circles', 'document__slug', 'document__title', 'document__license_url', 'ritual', 'concentration'],
        inplace=True)
open5e.rename(columns={'slug' : 'key', 'level_int' : 'level', 'dnd_class' : 'class'}, inplace=True)
open5e.columns

Index(['key', 'name', 'desc', 'higher_level', 'range', 'components',
       'duration', 'casting_time', 'level', 'school', 'class'],
      dtype='object')

Displaying the cleaned data

In [314]:
open5e.head()

key         name  \
0   acid-arrow   Acid Arrow   
1  acid-splash  Acid Splash   
2          aid          Aid   
3        alarm        Alarm   
4   alter-self   Alter Self   

                                                desc  \
0  A shimmering green arrow streaks toward a targ...   
1  You hurl a bubble of acid. Choose one creature...   
2  Your spell bolsters your allies with toughness...   
3  You set an alarm against unwanted intrusion. C...   
4  You assume a different form. When you cast the...   

                                        higher_level    range components  \
0  When you cast this spell using a spell slot of...  90 feet    V, S, M   
1                                                     60 feet       V, S   
2  When you cast this spell using a spell slot of...  30 feet    V, S, M   
3                                                     30 feet    V, S, M   
4                                                        Self       V, S   

        duration casting_time  level         school  \
0  Instantaneous     1 action      2      Evocation   
1  Instantaneous     1 action      0    Conjuration   
2        8 hours     1 action      2     Abjuration   
3        8 hours     1 minute      1     Abjuration   
4   Up to 1 hour     1 action      2  Transmutation   

                           class  
0                  Druid, Wizard  
1               Sorcerer, Wizard  
2                Cleric, Paladin  
3  Ranger, Ritual Caster, Wizard  
4               Sorcerer, Wizard

The initial data does not contain all of the inforamtion that I am looking for. I found the data in a different api, which is more difficult to work with.

In this data source, information on each spells will require a seperate api pull. This initial dataframe contains the names of each spell, along with the sub domains of the urls needed to pull the apis.

Data source: https://www.dnd5eapi.co/ 

In [315]:
r = requests.get('http://www.dnd5eapi.co/api/spells')
b = r.json()['results']
url_list = pd.DataFrame(data=b)
url_list.head()

index         name                      url
0   acid-arrow   Acid Arrow   /api/spells/acid-arrow
1  acid-splash  Acid Splash  /api/spells/acid-splash
2          aid          Aid          /api/spells/aid
3        alarm        Alarm        /api/spells/alarm
4   alter-self   Alter Self   /api/spells/alter-self

Counting the number of entries. 319 entries, matching the data from the other source.

In [316]:
url_list.count()

index    319
name     319
url      319
dtype: int64

I need to append the provided url subdirectories to the top level domain so I can run all of the apis. Here, I concatonate them and create a list.

In [317]:
spells_url = []

for url in url_list["url"]:
    spells_url.append('http://www.dnd5eapi.co' + url)

print(spells_url[0:5])

['http://www.dnd5eapi.co/api/spells/acid-arrow', 'http://www.dnd5eapi.co/api/spells/acid-splash', 'http://www.dnd5eapi.co/api/spells/aid', 'http://www.dnd5eapi.co/api/spells/alarm', 'http://www.dnd5eapi.co/api/spells/alter-self']


Reading in api for each spell, then concatonating them into a single dataframe.

In [318]:
full = []

for i in spells_url:
    data = requests.get(i)
    data_json = data.json()
    data_normal = pd.json_normalize(data_json)
    full.append(data_normal)
    

dnd5eapi = pd.concat(full)

In [319]:
dnd5eapi.reset_index(inplace=True)

In [320]:
dnd5eapi.count()

level_0                                319
index                                  319
name                                   319
desc                                   319
higher_level                           319
range                                  319
components                             319
material                               184
ritual                                 319
duration                               319
concentration                          319
casting_time                           319
level                                  319
attack_type                             16
classes                                319
subclasses                             319
url                                    319
damage.damage_type.index                64
damage.damage_type.name                 64
damage.damage_type.url                  64
damage.damage_at_slot_level.2           14
damage.damage_at_slot_level.3           17
damage.damage_at_slot_level.4           27
damage.dama

Displaying the data from dnd5eapi.co

In [321]:
dnd5eapi.head()

level_0        index         name  \
0        0   acid-arrow   Acid Arrow   
1        0  acid-splash  Acid Splash   
2        0          aid          Aid   
3        0        alarm        Alarm   
4        0   alter-self   Alter Self   

                                                desc  \
0  [A shimmering green arrow streaks toward a tar...   
1  [You hurl a bubble of acid. Choose one creatur...   
2  [Your spell bolsters your allies with toughnes...   
3  [You set an alarm against unwanted intrusion. ...   
4  [You assume a different form. When you cast th...   

                                        higher_level    range components  \
0  [When you cast this spell using a spell slot o...  90 feet  [V, S, M]   
1                                                 []  60 feet     [V, S]   
2  [When you cast this spell using a spell slot o...  30 feet  [V, S, M]   
3                                                 []  30 feet  [V, S, M]   
4                                                 []     Self     [V, S]   

                                        material  ritual       duration  \
0  Powdered rhubarb leaf and an adder's stomach.   False  Instantaneous   
1                                            NaN   False  Instantaneous   
2                   A tiny strip of white cloth.   False        8 hours   
3   A tiny bell and a piece of fine silver wire.    True        8 hours   
4                                            NaN   False   Up to 1 hour   

   concentration casting_time  level attack_type  \
0          False     1 action      2      ranged   
1          False     1 action      0         NaN   
2          False     1 action      2         NaN   
3          False     1 minute      1         NaN   
4           True     1 action      2         NaN   

                                             classes  \
0  [{'index': 'wizard', 'name': 'Wizard', 'url': ...   
1  [{'index': 'sorcerer', 'name': 'Sorcerer', 'ur...   
2  [{'index': 'cleric', 'name': 'Cleric', 'url': ...   
3  [{'index': 'ranger', 'name': 'Ranger', 'url': ...   
4  [{'index': 'sorcerer', 'name': 'Sorcerer', 'ur...   

                                          subclasses                      url  \
0  [{'index': 'lore', 'name': 'Lore', 'url': '/ap...   /api/spells/acid-arrow   
1  [{'index': 'lore', 'name': 'Lore', 'url': '/ap...  /api/spells/acid-splash   
2  [{'index': 'lore', 'name': 'Lore', 'url': '/ap...          /api/spells/aid   
3  [{'index': 'lore', 'name': 'Lore', 'url': '/ap...        /api/spells/alarm   
4  [{'index': 'lore', 'name': 'Lore', 'url': '/ap...   /api/spells/alter-self   

  damage.damage_type.index damage.damage_type.name  damage.damage_type.url  \
0                     acid                    Acid  /api/damage-types/acid   
1                     acid                    Acid  /api/damage-types/acid   
2                      NaN                     NaN                     NaN   
3                      NaN                     NaN                     NaN   
4                      NaN                     NaN                     NaN   

  damage.damage_at_slot_level.2 damage.damage_at_slot_level.3  \
0                           4d4                           5d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.4 damage.damage_at_slot_level.5  \
0                           6d4                           7d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.6 damage.damage_at_slot_level.7  \
0                           8d4                           9d4   

Columns from the dnd5eapi data. Note that many of the columns are redundant with the open5e data. 

In [322]:
dnd5eapi.columns

Index(['level_0', 'index', 'name', 'desc', 'higher_level', 'range',
       'components', 'material', 'ritual', 'duration', 'concentration',
       'casting_time', 'level', 'attack_type', 'classes', 'subclasses', 'url',
       'damage.damage_type.index', 'damage.damage_type.name',
       'damage.damage_type.url', 'damage.damage_at_slot_level.2',
       'damage.damage_at_slot_level.3', 'damage.damage_at_slot_level.4',
       'damage.damage_at_slot_level.5', 'damage.damage_at_slot_level.6',
       'damage.damage_at_slot_level.7', 'damage.damage_at_slot_level.8',
       'damage.damage_at_slot_level.9', 'school.index', 'school.name',
       'school.url', 'damage.damage_at_character_level.1',
       'damage.damage_at_character_level.5',
       'damage.damage_at_character_level.11',
       'damage.damage_at_character_level.17', 'dc.dc_type.index',
       'dc.dc_type.name', 'dc.dc_type.url', 'dc.dc_success',
       'heal_at_slot_level.2', 'heal_at_slot_level.3', 'heal_at_slot_level.4',
       

Removing unneeded columns and redundant columns, and renaming some other columns to suit my preference.

In [323]:
dnd5eapi.drop(columns=['level_0','name', 'desc', 'higher_level', 'range', 'components', 'material', 'duration', 'casting_time', 'level', 
                       'attack_type', 'classes', 'subclasses', 'url', 'damage.damage_type.index', 'damage.damage_type.url', 'school.index', 'school.name',
                       'school.url', 'dc.dc_type.index', 'dc.dc_type.name', 'dc.dc_type.url', 'dc.dc_success', 'dc.desc'],
        inplace=True)
dnd5eapi.rename(columns={'index':'key', 'damage.damage_type.name':'damage_type'}, inplace=True)
dnd5eapi.columns

Index(['key', 'ritual', 'concentration', 'damage_type',
       'damage.damage_at_slot_level.2', 'damage.damage_at_slot_level.3',
       'damage.damage_at_slot_level.4', 'damage.damage_at_slot_level.5',
       'damage.damage_at_slot_level.6', 'damage.damage_at_slot_level.7',
       'damage.damage_at_slot_level.8', 'damage.damage_at_slot_level.9',
       'damage.damage_at_character_level.1',
       'damage.damage_at_character_level.5',
       'damage.damage_at_character_level.11',
       'damage.damage_at_character_level.17', 'heal_at_slot_level.2',
       'heal_at_slot_level.3', 'heal_at_slot_level.4', 'heal_at_slot_level.5',
       'heal_at_slot_level.6', 'heal_at_slot_level.7', 'heal_at_slot_level.8',
       'heal_at_slot_level.9', 'area_of_effect.type', 'area_of_effect.size',
       'damage.damage_at_slot_level.1', 'heal_at_slot_level.1'],
      dtype='object')

Displaying cleaned data

In [324]:
dnd5eapi.head()

key  ritual  concentration damage_type  \
0   acid-arrow   False          False        Acid   
1  acid-splash   False          False        Acid   
2          aid   False          False         NaN   
3        alarm    True          False         NaN   
4   alter-self   False           True         NaN   

  damage.damage_at_slot_level.2 damage.damage_at_slot_level.3  \
0                           4d4                           5d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.4 damage.damage_at_slot_level.5  \
0                           6d4                           7d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.6 damage.damage_at_slot_level.7  \
0                           8d4                           9d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.8 damage.damage_at_slot_level.9  \
0                          10d4                          11d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_character_level.1 damage.damage_at_character_level.5  \
0                                NaN                                NaN   
1                                1d6                                2d6   
2                                NaN                                NaN   
3                                NaN                                NaN   
4                                NaN                                NaN   

  damage.damage_at_character_level.11 damage.damage_at_character_level.17  \
0                                 NaN                                 NaN   
1                                 3d6                                 4d6   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

  heal_at_slot_level.2 heal_at_slot_level.3 heal_at_slot_level.4  \
0                  NaN                  NaN                  NaN   
1                  NaN                  NaN                  NaN   
2                    5                   10                   15   
3                  NaN                  NaN                  NaN   
4                  NaN                  NaN                  NaN   

  heal_at_slot_level.5 heal_at_slot_level.6 heal_at_slot_level.7  \
0                  NaN                  NaN                  NaN   
1                  NaN                  NaN                  NaN   
2                   20                   25                   30   
3                  NaN                  NaN                  NaN   
4                  NaN                  NaN                  NaN   

  heal_at_slot_level.8 heal_at_slot_level.9 area_of_effect.type  \
0                  NaN                  NaN                 NaN   
1                  NaN                  NaN                 NaN   
2                   35                   40                 NaN   
3                  NaN                  NaN                cube   
4                  NaN                  NaN                 N

Merging the two dataframes. All of the data I was looking for is now contained in a single dataframe.

In [325]:
spells = pd.merge(open5e, dnd5eapi, on='key', how='left')
spells.head()

key         name  \
0   acid-arrow   Acid Arrow   
1  acid-splash  Acid Splash   
2          aid          Aid   
3        alarm        Alarm   
4   alter-self   Alter Self   

                                                desc  \
0  A shimmering green arrow streaks toward a targ...   
1  You hurl a bubble of acid. Choose one creature...   
2  Your spell bolsters your allies with toughness...   
3  You set an alarm against unwanted intrusion. C...   
4  You assume a different form. When you cast the...   

                                        higher_level    range components  \
0  When you cast this spell using a spell slot of...  90 feet    V, S, M   
1                                                     60 feet       V, S   
2  When you cast this spell using a spell slot of...  30 feet    V, S, M   
3                                                     30 feet    V, S, M   
4                                                        Self       V, S   

        duration casting_time  level         school  \
0  Instantaneous     1 action      2      Evocation   
1  Instantaneous     1 action      0    Conjuration   
2        8 hours     1 action      2     Abjuration   
3        8 hours     1 minute      1     Abjuration   
4   Up to 1 hour     1 action      2  Transmutation   

                           class ritual concentration damage_type  \
0                  Druid, Wizard  False         False        Acid   
1               Sorcerer, Wizard  False         False        Acid   
2                Cleric, Paladin  False         False         NaN   
3  Ranger, Ritual Caster, Wizard   True         False         NaN   
4               Sorcerer, Wizard  False          True         NaN   

  damage.damage_at_slot_level.2 damage.damage_at_slot_level.3  \
0                           4d4                           5d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.4 damage.damage_at_slot_level.5  \
0                           6d4                           7d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.6 damage.damage_at_slot_level.7  \
0                           8d4                           9d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_slot_level.8 damage.damage_at_slot_level.9  \
0                          10d4                          11d4   
1                           NaN                           NaN   
2                           NaN                           NaN   
3                           NaN                           NaN   
4                           NaN                           NaN   

  damage.damage_at_character_level.1 damage.damage_at_character_level.5  \
0                                NaN                                NaN   
1                                1d6                                2d6   
2                                NaN                                NaN   
3                                NaN                                NaN   
4                                NaN                                NaN   

  damage.damage_at_character_level.11 damage.damage_at_character_level.17  \
0                                 NaN                                 NaN   
1                                 3d6                                 4d6   
2                                 NaN 

Count of the merged dataframe. There are 319 entries, as expected.

In [326]:
spells.count()

key                                    319
name                                   319
desc                                   319
higher_level                           319
range                                  319
components                             319
duration                               319
casting_time                           319
level                                  319
school                                 319
class                                  319
ritual                                 316
concentration                          316
damage_type                             64
damage.damage_at_slot_level.2           14
damage.damage_at_slot_level.3           17
damage.damage_at_slot_level.4           27
damage.damage_at_slot_level.5           23
damage.damage_at_slot_level.6           32
damage.damage_at_slot_level.7           30
damage.damage_at_slot_level.8           30
damage.damage_at_slot_level.9           28
damage.damage_at_character_level.1      10
damage.dama

I noticed that there 3 NaN values in the "ritual" and "concentration" columns. This should not be that case. 

In [327]:
spells['ritual'].value_counts(dropna=False)

False    287
True      29
NaN        3
Name: ritual, dtype: int64

Finding the rows with NaN in "ritual". These rows also have NaN in "concentration".

In [328]:
spells[spells['ritual'].isna()]

key                name  \
12  antipathysympathy  Antipathy/Sympathy   
30  blindnessdeafness  Blindness/Deafness   
97      enlargereduce      Enlarge/Reduce   

                                                 desc  \
12  This spell attracts or repels creatures of you...   
30  You can blind or deafen a foe. Choose one crea...   
97  You cause a creature or an object you can see ...   

                                         higher_level    range components  \
12                                                     60 feet    V, S, M   
30  When you cast this spell using a spell slot of...  30 feet          V   
97                                                     30 feet    V, S, M   

          duration casting_time  level         school  \
12         10 days       1 hour      8    Enchantment   
30        1 minute     1 action      2     Necromancy   
97  Up to 1 minute     1 action      2  Transmutation   

                                      class ritual concentration damage_type  \
12                            Druid, Wizard    NaN           NaN         NaN   
30  Bard, Cleric, Sorcerer, Warlock, Wizard    NaN           NaN         NaN   
97                         Sorcerer, Wizard    NaN           NaN         NaN   

   damage.damage_at_slot_level.2 damage.damage_at_slot_level.3  \
12                           NaN                           NaN   
30                           NaN                           NaN   
97                           NaN                           NaN   

   damage.damage_at_slot_level.4 damage.damage_at_slot_level.5  \
12                           NaN                           NaN   
30                           NaN                           NaN   
97                           NaN                           NaN   

   damage.damage_at_slot_level.6 damage.damage_at_slot_level.7  \
12                           NaN                           NaN   
30                           NaN                           NaN   
97                           NaN                           NaN   

   damage.damage_at_slot_level.8 damage.damage_at_slot_level.9  \
12                           NaN                           NaN   
30                           NaN                           NaN   
97                           NaN                           NaN   

   damage.damage_at_character_level.1 damage.damage_at_character_level.5  \
12                                NaN                                NaN   
30                                NaN                                NaN   
97                                NaN                                NaN   

   damage.damage_at_character_level.11 damage.damage_at_character_level.17  \
12                                 NaN                                 NaN   
30                                 NaN                                 NaN   
97                                 NaN                                 NaN   

   heal_at_slot_level.2 heal_at_slot_level.3 heal_at_slot_level.4  \
12                  NaN                  NaN                  NaN   
30                  NaN                  NaN                  NaN   
97                  NaN                  NaN                  NaN   

   heal_at_slot_level.5 heal_at_slot_level.6 heal_at_slot_level.7  \
12                  NaN                  NaN                  NaN   
30                  NaN                  NaN                  NaN   
97                  NaN                  NaN                  NaN   

   heal_at_slot_level.8 heal_at_slot_level.9 area_of_effect.type  \
12                  NaN                  NaN                 NaN   
30                  NaN                  NaN                 NaN   
97                  NaN                  NaN                 NaN   

    area_of_effect.size damage.damage_at_slot_level.1 heal_at_slot_level.1  
12                  NaN                           NaN                  NaN  
30                  NaN                           NaN                  NaN  
97           

These spells have unique characteristics, so their data was listed incorrectly. I know that for all of these spells, "ritual" and "concentration" should be False.

Correcting this data.

In [329]:
spells.loc[[12]]

key                name  \
12  antipathysympathy  Antipathy/Sympathy   

                                                 desc higher_level    range  \
12  This spell attracts or repels creatures of you...               60 feet   

   components duration casting_time  level       school          class ritual  \
12    V, S, M  10 days       1 hour      8  Enchantment  Druid, Wizard    NaN   

   concentration damage_type damage.damage_at_slot_level.2  \
12           NaN         NaN                           NaN   

   damage.damage_at_slot_level.3 damage.damage_at_slot_level.4  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.5 damage.damage_at_slot_level.6  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.7 damage.damage_at_slot_level.8  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.9 damage.damage_at_character_level.1  \
12                           NaN                                NaN   

   damage.damage_at_character_level.5 damage.damage_at_character_level.11  \
12                                NaN                                 NaN   

   damage.damage_at_character_level.17 heal_at_slot_level.2  \
12                                 NaN                  NaN   

   heal_at_slot_level.3 heal_at_slot_level.4 heal_at_slot_level.5  \
12                  NaN                  NaN                  NaN   

   heal_at_slot_level.6 heal_at_slot_level.7 heal_at_slot_level.8  \
12                  NaN                  NaN                  NaN   

   heal_at_slot_level.9 area_of_effect.type  area_of_effect.size  \
12                  NaN                 NaN                  NaN   

   damage.damage_at_slot_level.1 heal_at_slot_level.1  
12                           NaN                  NaN

In [336]:
spells['ritual'] = spells['ritual'].fillna(False)
spells['concentration'] = spells['concentration'].fillna(False)

Verifying that the corrction has worked.

In [337]:
spells.loc[[12]]

key                name  \
12  antipathysympathy  Antipathy/Sympathy   

                                                 desc higher_level    range  \
12  This spell attracts or repels creatures of you...               60 feet   

   components duration casting_time  level       school          class  \
12    V, S, M  10 days       1 hour      8  Enchantment  Druid, Wizard   

    ritual  concentration damage_type damage.damage_at_slot_level.2  \
12   False          False         NaN                           NaN   

   damage.damage_at_slot_level.3 damage.damage_at_slot_level.4  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.5 damage.damage_at_slot_level.6  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.7 damage.damage_at_slot_level.8  \
12                           NaN                           NaN   

   damage.damage_at_slot_level.9 damage.damage_at_character_level.1  \
12                           NaN                                NaN   

   damage.damage_at_character_level.5 damage.damage_at_character_level.11  \
12                                NaN                                 NaN   

   damage.damage_at_character_level.17 heal_at_slot_level.2  \
12                                 NaN                  NaN   

   heal_at_slot_level.3 heal_at_slot_level.4 heal_at_slot_level.5  \
12                  NaN                  NaN                  NaN   

   heal_at_slot_level.6 heal_at_slot_level.7 heal_at_slot_level.8  \
12                  NaN                  NaN                  NaN   

   heal_at_slot_level.9 area_of_effect.type  area_of_effect.size  \
12                  NaN                 NaN                  NaN   

   damage.damage_at_slot_level.1 heal_at_slot_level.1  
12                           NaN                  NaN

In [339]:
spells['ritual'].value_counts(dropna=False)

False    290
True      29
Name: ritual, dtype: int64

In [340]:
spells['concentration'].value_counts(dropna=False)

False    194
True     125
Name: concentration, dtype: int64

In [341]:
spells.count()

key                                    319
name                                   319
desc                                   319
higher_level                           319
range                                  319
components                             319
duration                               319
casting_time                           319
level                                  319
school                                 319
class                                  319
ritual                                 319
concentration                          319
damage_type                             64
damage.damage_at_slot_level.2           14
damage.damage_at_slot_level.3           17
damage.damage_at_slot_level.4           27
damage.damage_at_slot_level.5           23
damage.damage_at_slot_level.6           32
damage.damage_at_slot_level.7           30
damage.damage_at_slot_level.8           30
damage.damage_at_slot_level.9           28
damage.damage_at_character_level.1      10
damage.dama